crates/astria-proto/src/native/sequencer/ics20_withdrawal.rs

```diff
@@ -0,0 +1,234 @@
+            timeout_height: Some(self.timeout_height.into()),
+            timeout_time: self.timeout_time,
+            source_channel: self.source_channel.to_string(),
+        }
+    }
+
+    /// Convert from a raw, unchecked protobuf [`raw::Ics20Withdrawal`].
+    ///
+    /// # Errors
+    ///
+    /// - if the `amount` field is missing
+    /// - if the `denom` field is invalid
+    /// - if the `return_address` field is invalid
+    /// - if the `timeout_height` field is missing
+    /// - if the `source_channel` field is invalid
+    pub fn try_from_raw(proto: raw::Ics20Withdrawal) -> Result<Self, Ics20WithdrawalError> {
+        let amount: Uint128 = proto.amount.ok_or(Ics20WithdrawalError::MissingAmount)?;
+        let return_address = Address::try_from_slice(&proto.return_address)
+            .map_err(Ics20WithdrawalError::InvalidReturnAddress)?;
+        let timeout_height = proto
+            .timeout_height
+            .ok_or(Ics20WithdrawalError::MissingTimeoutHeight)?
+            .into();
+
+        Ok(Self {
+            amount: amount.into(),
+            denom: IbcAsset::from_str(&proto.denom).map_err(Ics20WithdrawalError::InvalidDenom)?,
+            destination_chain_address: proto.destination_chain_address,
+            return_address,
+            timeout_height,
+            timeout_time: proto.timeout_time,
+            source_channel: ChannelId::from_str(&proto.source_channel)
+                .map_err(Ics20WithdrawalError::InvalidSourceChannel)?,
+        })
+    }
+}
+
+impl From<raw::IbcHeight> for IbcHeight {
+    fn from(h: raw::IbcHeight) -> Self {
+        Self {
+            revision_number: h.revision_number,
+            revision_height: h.revision_height,
+        }
+    }
+}
+
+impl crate::native::Protobuf for IbcHeight {
+    type Error = ::std::convert::Infallible;
+    type Raw = raw::IbcHeight;
+
+    fn try_from_raw_ref(raw: &Self::Raw) -> Result<Self, Self::Error> {
+        Ok(Self {
+            revision_number: raw.revision_number,
+            revision_height: raw.revision_height,
+        })
+    }
+
+    fn try_from_raw(h: Self::Raw) -> Result<Self, Self::Error> {
+        Ok(Self {
+            revision_number: h.revision_number,
+            revision_height: h.revision_height,
+        })
+    }
+
+    fn to_raw(&self) -> Self::Raw {
+        Self::Raw {
+            revision_number: self.revision_number,
+            revision_height: self.revision_height,
+        }
+    }
+
+    fn into_raw(self) -> Self::Raw {
+        Self::Raw {
+            revision_number: self.revision_number,
+            revision_height: self.revision_height,
+        }
+    }
+}
+
+impl From<IbcHeight> for raw::IbcHeight {
+    fn from(h: IbcHeight) -> Self {
+        Self {
+            revision_number: h.revision_number,
+            revision_height: h.revision_height,
+        }
+    }
+}
+
+#[allow(clippy::module_name_repetitions)]
+#[derive(Debug)]
+pub enum Ics20WithdrawalError {
```

Comment:
See how this is done in PR #646: derive `thiserror::Error` for this and split into a public wrapper type that has `#[error(transparent)]` and a private error type that has all the enums.